common-api/src/main/scala/pl/touk/nussknacker/engine/api/MetaData.scala

```diff
@@ -3,59 +3,106 @@ package pl.touk.nussknacker.engine.api
+      case "LiteStreamMetaData" => LiteStreamMetaData(additionalFields.properties)
+      case "RequestResponseMetaData" => RequestResponseMetaData(additionalFields.properties)
+      case "FragmentSpecificData" => FragmentSpecificData(additionalFields.properties)
+      case _ => throw new IllegalStateException("Unrecognized metadata type.")
+    }
+  }
 
-@ConfiguredJsonCodec sealed trait TypeSpecificData {
-  val isSubprocess = this match {
-    case _: ScenarioSpecificData => false
-    case _: FragmentSpecificData => true
+  def withTypeSpecificData(typeSpecificData: TypeSpecificData): MetaData = {
+    MetaData(id, typeSpecificData)
   }
 }
 
-sealed trait ScenarioSpecificData extends TypeSpecificData
-
-case class FragmentSpecificData(docsUrl: Option[String] = None) extends TypeSpecificData
+object MetaData {
+
+  private val actualDecoder: Decoder[MetaData] = deriveConfiguredDecoder[MetaData]
+
+  val legacyDecoder: Decoder[MetaData] = {
+    def legacyProcessAdditionalFieldsDecoder(scenarioType: String): Decoder[ProcessAdditionalFields] =
+      (c: HCursor) => for {
+        id <- c.downField("description").as[Option[String]]
+        properties <- c.downField("properties").as[Option[Map[String, String]]]
+      } yield {
+        ProcessAdditionalFields(id, properties.getOrElse(Map.empty), scenarioType)
+      }
+
+    (c: HCursor) => for {
+      id <- c.downField("id").as[String]
+      typeSpecificData <- c.downField("typeSpecificData").as[TypeSpecificData]
+      additionalFields <- c.downField("additionalFields")
+        .as[Option[ProcessAdditionalFields]](
+          io.circe.Decoder.decodeOption(
+            legacyProcessAdditionalFieldsDecoder(typeSpecificData.metaDataType)
+          )
+        ).map(_.getOrElse(ProcessAdditionalFields.empty(typeSpecificData.metaDataType)))
+    } yield {
+      MetaData(id, typeSpecificData, additionalFields)
+    }
+  }
 
-// TODO: rename to FlinkStreamMetaData
-case class StreamMetaData(parallelism: Option[Int] = None,
-                          //we assume it's safer to spill state to disk and fix performance than to fix heap problems...
-                          spillStateToDisk: Option[Boolean] = Some(true),
-                          useAsyncInterpretation: Option[Boolean] = None,
-                          checkpointIntervalInSeconds: Option[Long] = None) extends ScenarioSpecificData {
+  implicit val decoder: Decoder[MetaData] = actualDecoder or legacyDecoder
 
-  def checkpointIntervalDuration  : Option[Duration]= checkpointIntervalInSeconds.map(Duration.apply(_, TimeUnit.SECONDS))
+  def apply(id: String, typeSpecificData: TypeSpecificData, additionalFields: ProcessAdditionalFields): MetaData = {
+    MetaData(id = id, additionalFields = additionalFields.copy(
+      properties = additionalFields.properties ++ typeSpecificData.toMap
+    ))
+  }
 
+  def apply(id: String, typeSpecificData: TypeSpecificData): MetaData = {
```

Comment:
MetaData is constructed many times in tests (about 80 times). This complexity of having 2 legacy methods is for not having to change that many tests. This is because we want to handle 2 cases: 
- creating simple MetaData just from TypeSpecificData
- creating MetaData from ProcessAdditionalFields with TypeSpecificData - this is more complex because we merge these two into one generic map 

I can make one method from these two into something like this:
```scala
  def apply(id: String,
            typeSpecificData: TypeSpecificData,
            additionalFields: Option[ProcessAdditionalFields] = None): MetaData = {
    MetaData(
      id = id,
      additionalFields = ProcessAdditionalFields(
        additionalFields.flatMap(_.description),
        typeSpecificData.toMap ++ additionalFields.map(_.properties).getOrElse(Map.empty),
        typeSpecificData.metaDataType
      )
    )
  }
```
Then we can delete that combineTypeSpecificProperties method. Also this should probably be in a factory method, not in constructor, but then it will require changing all the tests. WDYT?